### ISSUE TITLE
Incorrect Literal type widening in method calls

### ISSUE THREAD

**Issue by tjf801** [microsoft/pyright#4890]
**Describe the bug**
Generic `Literal` integer types get widened to `int` in certain methods when they should stay narrowed.

**Expected behavior**
Calling methods with types, and explicitly showing the return annotation should not widen Literal types to Int.

**Screenshots or Code**
```python
from typing import Generic, Literal, TypeVar

_N = TypeVar('_N')
_M = TypeVar('_M')

class A(Generic[_M, _N]):
    def __new__(
        cls,
        m: _M,
        n: _N,
    ) -> A[_M, _N]: ...

# Expression of type "A[int, int]" cannot be assigned to declared type "A[Literal[3], Literal[4]]"
a: A[Literal[3], Literal[4]] = A(3, 4)
```

**VS Code extension or command-line**
I'm using the Pyright VS Code Extension.

**Additional context**
After some experimentation, it seems that using a function with a tuple return type *does* type check, as in this example.
```python
from typing import Literal, TypeVar

_N = TypeVar('_N')
_M = TypeVar('_M')

def f(x: _N, y: _M) -> tuple[_N, _M]: ...

a: tuple[Literal[3], Literal[4]] = f(3, 4)
```
Similarly, setting `_N` and `_M` to be covariant had no effect on the type error.

**Comment by erictraut** [comment]
Type checking for a constructor is complicated because the type checker needs to consider both the `__init__` and `__new__` methods. Both of these can use class-scoped TypeVars and both can optionally be overloaded. Both can also be used with bidirectional type inference, as you are doing in your code example.

It's unusual to use a `__new__` method in the way you're doing here. Most classes would use an `__init__` method instead. If you change the code to use an `__init__` method, you will not hit this bug.

```python
class A(Generic[_M, _N]):
    def __init__(self, m: _M, n: _N) -> None:
        ...
```

That said, I agree this is a bug. The `__new__` method should work with bidirectional type inference in this case.

This will be addressed in the next release.

Incidentally, if you're going to stick with the `__new__` method, I recommend annotating its return type as `Self` rather than `A[_M, _N]` because the latter will not work with subclasses.

**Comment by tjf801** [comment]
That's probably the quickest I've ever had a bug report fixed. Thank you!

(And yeah, I know using `__new__` here is quite unusual, and I should have used `typing.Self`, but I found this problem in a simple stub file I was writing for an extension module. I made the above example with the simplest example I could.)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.302, which I just published. It will also be included in a future release of pylance.
